src/t_zset.c

```diff
@@ -328,54 +329,64 @@ int zslIsInRange(zskiplist *zsl, zrangespec *range) {
     return 1;
 }
 
-/* Find the first node that is contained in the specified range.
- * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslFirstInRange(zskiplist *zsl, zrangespec *range) {
+/* Find the Nth node that is contained in the specified range. N should be 0-based.
+ * Negative N works for reversed order (-1 represents the last element). Returns
+ * NULL when no element is contained in the range. */
+zskiplistNode *zslNthInRange(zskiplist *zsl, zrangespec *range, long n) {
     zskiplistNode *x;
     int i;
+    long rank = 0;
 
     /* If everything is out of range, return early. */
     if (!zslIsInRange(zsl,range)) return NULL;
 
     x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *OUT* of range. */
-        while (x->level[i].forward &&
-            !zslValueGteMin(x->level[i].forward->score,range))
+    if (n >= 0) {
+        for (i = zsl->level - 1; i >= 0; i--) {
+            /* Go forward while *OUT* of range. */
+            while (x->level[i].forward && !zslValueGteMin(x->level[i].forward->score, range)) {
+                /* Count the rank of the first element in range. */
+                rank += x->level[i].span;
                 x = x->level[i].forward;
-    }
-
-    /* This is an inner range, so the next node cannot be NULL. */
-    x = x->level[0].forward;
-    serverAssert(x != NULL);
-
-    /* Check if score <= max. */
-    if (!zslValueLteMax(x->score,range)) return NULL;
-    return x;
-}
-
-/* Find the last node that is contained in the specified range.
- * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslLastInRange(zskiplist *zsl, zrangespec *range) {
-    zskiplistNode *x;
-    int i;
-
-    /* If everything is out of range, return early. */
-    if (!zslIsInRange(zsl,range)) return NULL;
-
-    x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *IN* range. */
-        while (x->level[i].forward &&
-            zslValueLteMax(x->level[i].forward->score,range))
+            }
+        }
+        /* Check if zsl is long enough. */
+        if ((unsigned long)(rank + n) >= zsl->length) return NULL; 
+        if (n < ZSKIPLIST_MAX_SEARCH) {
+            /* If offset is small, we can just jump node by node */
+            for (i = 0; i < n + 1; i++) { 
+                x = x->level[0].forward; 
+            }
+        } else {
+            /* If offset is big, we jump to the nth node by its rank from the start of skiplist. */
+            x = zslGetElementByRank(zsl, rank + n + 1);
```

Comment:
i don't think i understand why we can't always use zslGetElementByRank? i.e. why do we need ZSKIPLIST_MAX_SEARCH and the other branch of that if-else?
just to avoid going through all levels?
maybe instead, an optimization in zslGetElementByRank is due?

Reply:
zslGetElementByRank can only start searching from the beginning of a zsl. Now that we already reached the `rank`th node, If `rank` is close enough to `n`, it is quicker to reach the target node by stepping forward from `rank`th node, than restart from the beginning.

Reply:
We can conduct further testing to explore the performance differences in different scenarios when the offset is set to 1/4/8/16/32/64.

Reply:
and can't we make a version of zslGetElementByRank that will be capable of both?
i.e. both using multiple levels, and also starting from a certain position rather than the start?
e.g. `zskiplistNode* zslGetElementByRank(zskiplist *zsl, unsigned long rank, zskiplistNode *search_from_node, long search_from_node_rank);`?

Reply:
Certainly we can.

But I am not sure how much improvement this optimizaiton can bring. And the only use case of `search_from_node` and `search_from_node_rank` is in `ZRANGE` command. I'd prefer to keep the if-else here.

Reply:
my question wasn't just about moving the if-else to zslGetElementByRank, but more about utilizing all levels even when we're searching from a certain point (instead having to choose between using just the first level, or searching from the head).

or maybe we can't do that when we start from a zskiplistNode (which can be at any level), and we must keep a pointer to the last `[zsl->level-1]` node we visited?
unless i'm missing something, that will add some complexity (not extremely high), but can let us enjoy both optimizations.

Reply:
or another approach (slightly more complicated, is to step forward on level 0 until we find a node with higher level, and continue stepping with the higher level from that point onward 
![image](https://github.com/redis/redis/assets/7045099/ee568fe6-77f3-43c8-bc6e-6dc88b59ef99)

Reply:
The problem is, currently we would not know the how many levels a zskiplistNode has unless we search for it from the head (`level[]` is a flexible array at the tail of `struct zskiplistNode`).  Unless we want to add a new member into `struct zskiplistNode`.

Reply:
ok, i wasn't aware of that. so can't zslNthInRange remember the last `[zsl->level-1]` node we visited and let zslGetElementByRank start from there?

Reply:
Yes and it saves a few steps than searching from the head.

Reply:
> Yes and it saves a few steps than searching from the head

it can save a lot, depending on the size of the skiplist and how far we are from the head.
IMHO we should do it.. not sure what's the cleanest way, modify zslGetElementByRank or create a new helper function, please look into it.

Reply:
ok I'll try.